store/cassandra/cassandra.go

```diff
@@ -395,30 +499,39 @@ type readResult struct {
 }
 
 func (c *CassandraStore) processReadQueue() {
-	for crr := range c.readQueue {
-		// check to see if the request has been canceled, if so abort now.
+	defer c.wg.Done()
+
+	for {
 		select {
-		case <-crr.ctx.Done():
-			//request canceled
-			crr.out <- readResult{err: errCtxCanceled}
-			continue
-		default:
-		}
-		waitDuration := time.Since(crr.timestamp)
-		cassGetWaitDuration.Value(waitDuration)
-		if waitDuration > c.omitReadTimeout {
-			cassOmitOldRead.Inc()
-			crr.out <- readResult{err: errReadTooOld}
-			continue
-		}
+		case crr := <-c.readQueue:
+			// check to see if the request has been canceled, if so abort now.
+			select {
+			case <-crr.ctx.Done():
+				//request canceled
+				crr.out <- readResult{err: errCtxCanceled}
+				continue
+			default:
+			}
+			waitDuration := time.Since(crr.timestamp)
+			cassGetWaitDuration.Value(waitDuration)
+			if waitDuration > c.omitReadTimeout {
+				cassOmitOldRead.Inc()
+				crr.out <- readResult{err: errReadTooOld}
+				continue
+			}
 
-		pre := time.Now()
-		iter := readResult{
-			i:   c.Session.Query(crr.q, crr.p...).WithContext(crr.ctx).Iter(),
-			err: nil,
+			// the sessionLocks are handled in SearchTable
+			pre := time.Now()
+			iter := readResult{
+				i:   c.Session.Query(crr.q, crr.p...).WithContext(crr.ctx).Iter(),
```

Comment:
i think that's missing the sessionLock

Reply:
Those locks are handled in SearchTable, which is the only function that accesses the readQueue

Reply:
IC, so yeah that works because we know that SearchTable is the only location which ever pushes jobs into the readQueue and it waits for the result before releasing the lock. However, that feels really fragile because it relies on the interactions between two queue producers / consumers to work exactly the way they currently do to guarantee that the use of `c.Session` is protected by the lock. If somebody in the future modifies those interactions it would be really easy to not realize that `processReadQueue` also needs to be holding the read lock. Since multiple routines can hold the read lock concurrently, i'd just make both `processReadQueue()` and `SearchTable()` acquire it separately. This shouldn't lead to any slow down because we know that in the current state of the code whenever `processReadQueue()` executes a job then `SearchTable()` is already holding the read lock, which means nothing is holding the write lock, so `processReadQueue()` should be able to acquire it without waiting.

Reply:
Adding locks `processReadQueue` does not benefit us. If the basic flow of this process changes in the future then it will require other things to change as well. There is already a note about where the session locks are handled, right before this block of code. I can expand on that comment and make it more clear that if anyone modifies the function that they will need to change how the locks are handled.

In its current state adding a lock would not help us because the calling function is still responsible for locking and unlocking the session. The session must remain locked by the caller until it is done processing the returned iterator / error. Adding locks to the read queue would be a false sense of security, because they don't actually prevent the session from expiring between the the producer sending the data to the channel and the consumer reading it and then trying to use the iterator. I think it is more dangerous to add locks to `processReadQueue`

Reply:
ok, fair enough. let's just make this really clear in the comments then, that `processReadQueue` needs to be lock protected and it currently is because `SearchTable()` holds the lock while waiting.